src/soc/ibm/power9/chip.c

```diff
@@ -254,105 +280,88 @@ static void fill_cpu_node(struct device_tree *tree,
 
-/*
- * coreboot prepares one node with multiple "reg" entries (see fit_update_memory
- * and functions called by it), while Skiboot expects multiple nodes, one region
- * each. Both approaches are compliant with Devicetree Specification v0.3 so we
- * can't say that one of them is wrong, they're just different.
- *
- * Probable cause for Skiboot choosing multi-node version is that it is possible
- * to specify associativity, which may be important for NUMA systems.
- *
- * TODO: create these nodes from DIMM entries, when they are implemented.
- */
-static void split_mem_node(struct device_tree *tree)
+static void add_memory_node(struct device_tree *tree, uint8_t chip, uint64_t start,
+			    uint64_t len)
 {
-	struct device_tree_node *old_node;
-	struct device_tree_node *new_region;
-	struct device_tree_property *prop;
-	const uint32_t *reg_data = NULL;
-	size_t count, size_bytes = 0, offset = 0;
-	/*
-	 * Assume that, unless specified otherwise in "memory" node, defaults of 2
-	 * are used for address and size cells.
-	 */
-	size_t address_cells = 2, size_cells = 2;
+	struct device_tree_node *node;
+	/* /memory@0123456789abcdef - 24 characters + null byte */
+	char path[26] = {};
 
-	old_node = dt_find_node_by_path(tree, "/memory", NULL, NULL, 0);
-	if (old_node == NULL)
-		die("No 'memory' node in device tree!\n");
+	union {uint32_t u32[2]; uint64_t u64;} addr = { .u64 = start };
+	union {uint32_t u32[2]; uint64_t u64;} size = { .u64 = len };
 
-	list_remove(&old_node->list_node);
+	snprintf(path, sizeof(path), "/memory@%llx", addr.u64);
+	node = dt_find_node_by_path(tree, path, NULL, NULL, 1);
 
-	list_for_each(prop, old_node->properties, list_node) {
-		if (!strcmp("reg", prop->prop.name)) {
-			size_bytes = prop->prop.size;
-			reg_data = (uint32_t *)prop->prop.data;
-			continue;
-		}
-		if (!strcmp("#address-cells", prop->prop.name)) {
-			address_cells = *(uint32_t *)prop->prop.data;
-			continue;
-		}
-		if (!strcmp("#size-cells", prop->prop.name)) {
-			size_cells = *(uint32_t *)prop->prop.data;
-			continue;
-		}
-	}
+	dt_add_string_prop(node, "device_type", (char *)"memory");
+	/* Use 2 cells each for address and size. This assumes BE. */
+	dt_add_reg_prop(node, &addr.u64, &size.u64, 1, 2, 2);
 
-	if (reg_data == NULL)
-		die("No 'reg' property in 'memory' node\n");
+	/* Don't know why the value needs to be shifted (group id + chip id?) */
+	dt_add_u32_prop(node, "ibm,chip-id", chip << 3);
+}
 
-	if (address_cells != 2 && address_cells != 1)
-		die("'#address_cells' property in 'memory' has invalid value (%d)\n",
-		    address_cells);
+static bool build_memory_map(const struct range_entry *r, void *arg)
+{
+	struct mem_map *map = arg;
 
-	if (size_cells != 2 && size_cells != 1)
-		die("'#size_cells' property in 'memory' has invalid value (%d)\n",
-		    size_cells);
+	/*
+	 * Kernel likes its available memory areas at least 1MB
+	 * aligned, let's trim the regions such that unaligned padding
+	 * is added to reserved memory.
+	 */
```

Comment:
This probably applies to some embedded ARM platform, skiboot doesn't mind such regions. This whole function can be simplified to "if RAM then add to mem, else to reserved", without splicing ranges.

Reply:
Actually, this may work different than Hostboot, where "memory" nodes reported all RAM, reserved or not. Reading `MCS_MCFGP*` may be the only correct option. `bootmem_walk_os_mem` would be used just for reserved regions.